text/0041-fixed-point.md

```diff
@@ -0,0 +1,143 @@
+Without a first class fixed point type, the user has to manually manage how the value needs to be shifted to be represented as an integer and keep track of how that interacts with arithmetic operations.
+
+A fixed point type would encode and keep track of the precision through arithmetic operations, as well as provide standard operations for converting values to and from fixed point representations with correct rounding.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+TODO
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC proposes a library addition `amaranth.lib.fixed` with the following contents:
+
+`fixed.Shape` is a `ShapeCastable` subclass.
+The following operations are defined on it:
+
+- `fixed.Shape(f_width, /, *, signed)`: Create a `fixed.Shape` with zero integer bits and `f_width` fractional bits.
+- `fixed.Shape(i_width, f_width, /, *, signed)`: Create a `fixed.Shape` with `i_width` integer bits and `f_width` fractional bits.
+  - The sign bit is not included in `i_width` or `f_width`, so a `fixed.Shape(7, 8, signed=True)` will be 16 bits wide.
+- `fixed.Shape.cast(shape, f_width=0)`: Cast `shape` to a `fixed.Shape` instance.
+- `.i_width`, `.f_width`, `.signed`: Width and signedness properties.
+- `.const(value)`: Create a `fixed.Const` from `value`.
+- `.as_shape()`: Return the underlying `Shape`.
+- `.__call__(target)`: Create a `fixed.Value` over `target`.
+
+`SQ(*args)` is an alias for `fixed.Shape(*args, signed=True)`.
+
+`UQ(*args)` is an alias for `fixed.Shape(*args, signed=False)`.
+
+`fixed.Value` is a `ValueCastable` subclass.
+The following operations are defined on it:
+
+- `fixed.Value(shape, target)`: Create a `fixed.Value` with `shape` over `target`.
+- `fixed.Value.cast(value, f_width=0)`: Cast `value` to a `fixed.Value`.
+- `.i_width`, `.f_width`, `.signed`: Width and signedness properties.
+- `.shape()`: Return the `fixed.Shape` this was created from.
+- `.as_value()`: Return the underlying value.
+- `.eq(value)`: Assign `value`.
+  - If `value` is a `Value`, it'll be assigned directly to the underlying `Value`.
+  - If `value` is an `int` or `float`, it'll be cast to a `fixed.Const` first.
+  - If `value` is a `fixed.Value`, the precision will be extended or rounded as required.
+- `.round(f_width=0)`: Return a new `fixed.Value` with precision changed to `f_width`, rounding as required.
+- `.__add__(other)`, `.__radd__(other)`, `.__sub__(other)`, `.__rsub__(other)`, `.__mul__(other)`, `.__rmul__(other)`: Binary arithmetic operators.
+  - If `other` is a `Value`, it'll be cast to a `fixed.Value` first.
+  - If `other` is an `int`, it'll be cast to a `fixed.Const` first.
+  - If `other` is a `float`: TBD
+  - The result will be a new `fixed.Value` with enough precision to hold any resulting value without rounding or overflowing.
+- `.__lshift__(other)`, `.__rshift__(other)`: Bit shift operators.
```

Comment:
In the implementation at least the semantics of these seem to be different from underlying amaranth values. If the shift amount is an integer these act like `.shift_left()` and `.shift_right()`. Those operators should probably be added here and the semantics of `__lshift__` and `__rshift__` adjusted to match those on `unsigned()` and `signed()`.

Note: I may be wrong here, still learning the language